qiskit_experiments/stored_data/stored_data.py

```diff
@@ -0,0 +1,1047 @@
+                        except Exception:  # pylint: disable=broad-except
+                            pass
+                    if self._jobs[jid] is not None:
+                        self._add_result_data(self._jobs[jid].result())
+
+        if index is None:
+            return self._data.copy()
+        if isinstance(index, (int, slice)):
+            return self._data[index]
+        if isinstance(index, str):
+            return [data for data in self._data if data.get("job_id") == index]
+        raise TypeError(f"Invalid index type {type(index)}.")
+
+    @auto_save
+    def add_figures(
+        self,
+        figures: Union[List[Union[str, bytes, "pyplot.Figure"]], str, bytes, "pyplot.Figure"],
+        figure_names: Optional[Union[List[str], str]] = None,
+        overwrite: bool = False,
+        save_figure: Optional[bool] = None,
+        service: Optional["ExperimentServiceV1"] = None,
+    ) -> Union[str, List[str]]:
+        """Add the experiment figure.
+
+        Args:
+            figures: Names of the figure files or figure data.
+            figure_names: Names of the figures. If ``None``, use the figure file
+                names, if given, or a generated name. If `figures` is a list, then
+                `figure_names` must also be a list of the same length or ``None``.
+            overwrite: Whether to overwrite the figure if one already exists with
+                the same name.
+            save_figure: Whether to save the figure in the database. If ``None``,
+                the ``auto-save`` attribute is used.
+            service: Experiment service to be used to update the database, if
+                the figure is to be uploaded. If ``None``, the default service is used.
+
+        Returns:
+            Figure names.
+
+        Raises:
+            ExperimentEntryExists: If the figure with the same name already exists,
+                and `overwrite=True` is not specified.
+            ValueError: If an input parameter has an invalid value.
+        """
+        if (
+            isinstance(figures, list)
+            and figure_names is not None
+            and (not isinstance(figure_names, list) or len(figures) != len(figure_names))
+        ):
+            raise ValueError(
+                "The parameter figure_names must be None or a list of "
+                "the same size as the parameter figures."
+            )
+        if not isinstance(figures, list):
+            figures = [figures]
+        if figure_names is not None and not isinstance(figure_names, list):
+            figure_names = [figure_names]
+
+        added_figs = []
+        for idx, figure in enumerate(figures):
+            if figure_names is None:
+                if isinstance(figure, str):
+                    fig_name = figure
+                else:
+                    fig_name = (
+                        f"figure_{self.experiment_id[:8]}_"
+                        f"{datetime.now().isoformat()}_{len(self._figures)}"
```

Comment:
Sounds like we need to retrieve all figures anyways because we cannot guess what is drawn from the timestamp. Can you add some mechanism to pass image tag as I wrote in another comment?